SRCH-1470 - Add click tracking to API V2

ondrae opened the pull request:
## A new click tracking endpoint
This PR adds a new click tracking endpoint to the v2 api, customer documentation for that endpoint, and lots of tests.

It also changes how the existing click tracking from a serp page works. Much of the existing click tracking was untested, now it is covered. There is a new system spec intended to ensure the javascript for serp clicks works correctly. To support this test, we switched out poltergeist for headless chrome.

## Before deploying
- Logstash will need to be updated to read the new `[Click]` loglines, instead of the request loglines.  You'll know if it came from the api endpoint because it will have an `access_key` param included.
- Notify https://api.data.gov/ about the new endpoint.
- Update external docs at https://open.gsa.gov/api/searchgov-results/ to match the new internal docs for this endpoint.

### Deploying Notes
We changed existing click tracking from a GET to POST. After deploying, users with cached serps may have some unsuccessful requests for click tracking until they reload the page.

### What we learned
- You can't use the standard validators when checking against fixture supplied data. The validation is run before fixture data is loaded. [See comment here](https://github.com/GSA/search-gov/blob/bf8b05199a61589c0b0b056140013248ba5f7bfd/app/models/click.rb#L43).
- GSA laptops are slow to upgrade Chrome. The webdrivers gem will check for your installed version, to ensure it works.
- `after_validation` is a useful way to transform params after checking them.
- Note to self, being overly DRY in specs can take a lot of time and just needs to be undone when the design changes. 
- We could have split this into two separate PRs. One to update the current SERP click tracking. A second to add the new api endpoint.

### Not included in this PR
- Upgrading Capybara. We wanted to get rid of some deprecation warnings, yet upgrading causes some new specs to fail.
- Removing unused code tangentially related to click tracking. There are [existing stories](https://cm-jira.usa.gov/browse/SRCH-368) to do just that.

MothOnMars commented on app/controllers/clicked_controller.rb:
```diff
@@ -1,20 +1,31 @@
-      Click.log(url, query, queried_at_gmt, click_ip, affiliate_name, position, results_source, vertical, locale, user_agent, model_id)
-    end
-    self.response_body = ""
+    query = params['q']
+    position = params['p']
+    affiliate_name = params['a'].presence
+    results_source = params['s']
+    vertical = params['v']
+    click_ip = request.env['REMOTE_ADDR']
```
Let's chat re. parameter naming

ondrae commented on app/views/sites/click_tracking_api_instructions/_show_md.html.haml:
```diff
@@ -0,0 +1,31 @@
+  - `query`: The search term that let this result get shown and clicked on.
+  - `affiliate`: You can find your site handle on the [Settings](#{edit_site_setting_path(@site)}) page.
+  - `position`:  The position/rank on your search results page. Was it the first result or the second?
+  - `source`: The module code for the source of the clicked result. Must be a valid module code. See https://search.gov/manual/module-codes.html
```
- Use `module` instead of source.

ondrae commented on app/models/click.rb:
```diff
@@ -1,17 +1,15 @@
     click_hash = { :clicked_at => Time.now.to_formatted_s(:db),
                    :url => url,
-                   :queried_at => queried_at,
                    :affiliate_name => affiliate_name,
```
- Use `affiliate` here to mean the name.

ondrae commented on app/views/sites/click_tracking_api_instructions/_show_md.html.haml:
```diff
@@ -0,0 +1,31 @@
+
+  ## Optional Parameters
+  - `vertical`: The search engine your site uses.
+  - `click_ip`: The ip address of the user who clicked.
```
Change it everywhere to `client_ip`.

ondrae replied:
`client_ip` and `user_agent` should become required.

ondrae commented on spec/helpers/results_helper_spec.rb:
```diff
@@ -0,0 +1,35 @@
+      double('search',
+             affiliate: affiliates(:basic_affiliate),
+             query: 'rutabaga',
+             queried_at_seconds: "1588714504")
```
Don't need `queried_at_seconds` anymore.

ondrae replied:
Fixed this.

ondrae commented on app/views/sites/click_tracking_api_instructions/_show_md.html.haml:
```diff
@@ -0,0 +1,31 @@
+  **POST**
+
+  ## Required Parameters
+  - `clicked_url`: The url of the link that was clicked.
```
Use `url` instead of clicked_url

MothOnMars commented on app/helpers/results_helper.rb:
```diff
@@ -2,16 +2,14 @@ module ResultsHelper
         a: search.affiliate.name,
-        l: search.affiliate.locale,
-        q: search.query,
-        s: search.module_tag,
```
Per our chat, the module tag was being added twice

ondrae commented on app/controllers/api/v2/click_controller.rb:
```diff
@@ -0,0 +1,45 @@
+        query = params['query']
+        position = params['position']
+        affiliate = params['affiliate'].presence
+        source = params['source']
```
module is a reserved word in ruby. https://docs.ruby-lang.org/en/2.7.0/keywords_rdoc.html

ondrae replied:
`module_source`? `module_code`?

dawnpm replied:
`Source` really speaks to 'what index did these results come from?' as opposed to `module` which talks about which block in the results page did the result appear in. For the purpose of click tracking, we do report how the various blocks are performing - best bets vs rss vs web, etc. - currently modules/display blocks are tied to a single index, but someday I hope the `web` module will be multi-source (either Bing or i14y).

MothOnMars replied:
@dawnpm @ondrae , the `source` in terms of click tracking translates to the `modules` field in the logstash documents. So `module_code` is a good name for that field in the API (for clicks, only one module code should ever be used, so `modules` would not be accurate).

ondrae commented:
![Screen Shot 2020-06-03 at 2 54 09 PM](https://user-images.githubusercontent.com/595778/83693325-ba7d0400-a5aa-11ea-985b-0a0b2429ff0e.png)
https://github.com/GSA/search-gov/blob/47f7c08743a35ad38263d93278a96ba35fa3033a/app/views/sites/click_tracking_api_instructions/_show_md.html.haml

@MothOnMars @ayushchak The customer facing docs page is ready for you to make copywriting edits to. You can send me the edits you want to make, or set up a PR against my working branch.

A few questions:
- Whats a good way to describe the position parameter?
- Module is a reserved word in ruby. I used `module_code`, that okay?
- Is client_ip and user_agent used for anything besides bot detection?
- Do we want to let customers know about the optional vertical parameter?

MothOnMars commented:
Answers to your q's:
- Whats a good way to describe the position parameter?
See https://search.gov/manual/clicks.html

- Module is a reserved word in ruby. I used module_code, that okay?
Yep.

- Is client_ip and user_agent used for anything besides bot detection?
Yes, the client ip is used by Logstash for geolocation, so each click document in our logstash indices contains location data such as country code, region, etc. That data is available in the CSV files that admins can download for their click analytics.

- Do we want to let customers know about the optional vertical parameter?
That's a product q/ for @dawnpm or @ayushchak . We use `vertical` to identify the particular SERP type where the click originated (`blended`,`docs`, `web`, etc), but I don't know if our customers want that level of granularity to identify clicks coming from different types of results pages....? (Note: "The search engine your site uses" is not an accurate description of the `vertical`.)

MothOnMars commented on spec/models/click_api_spec.rb:
```diff
@@ -0,0 +1,82 @@
+    end
+
+    describe '#log' do
+      it 'should log almost-JSON info about the click' do
```
For this & the other specs, avoid using `should` in the description, i.e.:
`it 'logs almost-JSON info about the click' do`

http://www.betterspecs.org/#should

MothOnMars commented on spec/requests/api/v2/click_spec.rb:
```diff
@@ -0,0 +1,109 @@
+    end
+
+    it 'sends the expected params to Click' do
+      expect(ClickApi).to receive(:new).with(
+        access_key: 'basic_key',
+        affiliate: 'nps.gov',
+        client_ip: '127.0.0.1',
+        module_code: 'test_source',
+        position: '1',
+        query: 'test_query',
+        url: 'https://search.gov',
+        user_agent: 'test_user_agent',
+        vertical: 'test_vertical'
```
I would DRY this up: `expect(ClickApi).to receive(:new).with(params)`

MothOnMars commented on spec/requests/api/v2/click_spec.rb:
```diff
@@ -0,0 +1,109 @@
+
+      post '/api/v2/click', params: params
+      expect(response.success?).to be(false)
+      expect(response.body).to eq "[\"Url can't be blank\"]"
```
Hrm. This doesn't seem like a useful error in this scenario. Maybe we should change the `cleaned_url` method into a proper validation, and validate the URL, so that we can return something along the lines of `URL is not a valid format`?

MothOnMars commented on spec/requests/clicked_serp_spec.rb:
```diff
@@ -0,0 +1,81 @@
+require 'spec_helper'
+
+describe 'Clicked' do
```
We could DRY up the api vs. non-api request specs by using some shared behavior, i.e. `it_behaves_like 'a click request'`

ondrae commented:
@MothOnMars Thank you for the great review!

ondrae commented on app/models/click_api.rb:
```diff
@@ -0,0 +1,27 @@
+class ClickApi < ClickSerp
```
Naming things, ugh. I will go with the simplified `Click` and `ApiClick`.

MothOnMars replied:
👍

ondrae commented on app/models/click_serp.rb:
```diff
@@ -0,0 +1,28 @@
+  def initialize(params)
+    @url = cleaned_url(params[:url])
+    @query = params[:query]
+    @client_ip = params[:client_ip]
```
Should I validate ip6 too?

MothOnMars replied:
`Resolv::AddressRegex` *should* cover both formats:
https://www.rubydoc.info/github/ydakuka/ruby-ru/master/Resolv

ondrae commented on spec/helpers/results_helper_spec.rb:
```diff
@@ -0,0 +1,35 @@
+    let(:search_vertical) { 'i14y' }
+
+    it 'adds data attributes to #search needed for click tracking' do
+      output = search_data(search, search_vertical)
```
You can't name a subject the same as the underlying method, so I just went with
```
subject { search_data(search, search_vertical) }
...
is_expected.to eq expected_output
```

ondrae replied:
I like this a little more `expect(subject).to eq expected_output`

ondrae commented on app/views/sites/click_tracking_api_instructions/_show_md.html.haml:
```diff
@@ -0,0 +1,28 @@
+  ## Method
+  **POST**
+
+  ## Required Parameters
```
I've asked Ayush on Jira [here](https://cm-jira.usa.gov/browse/SRCH-1470?focusedCommentId=799969&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-799969)

ondrae replied:
They said to not include `referrer`.

ondrae commented on app/views/sites/click_tracking_api_instructions/_show_md.html.haml:
```diff
@@ -0,0 +1,28 @@
+  - `access_key`: Your [API access key](#{api_scheme_and_host}/sites/#{@site.id}/api_access_key).
+  - `client_ip`: The ip address of the user who clicked.
+  - `user_agent`: The user agent of the user who clicked.
+  
```
Asked Ayush on Jira.

ondrae replied:
Vertical will be a non-required param for Clicks. It will be undocumented for ApiClicks, but not explicitly blocked.

ondrae commented on spec/helpers/results_helper_spec.rb:
```diff
@@ -0,0 +1,35 @@
+
+  describe '#link_to_result_title' do
+    it 'makes a link with the added data-click attribute' do
+      output = link_to_result_title('test title', 'https://test.gov', '2', 'BOOS')
```
I've changed both of these to use subject. It usually feels more appropriate when it is used more than once though.

ondrae commented on spec/system/serp_click_tracking_spec.rb:
```diff
@@ -0,0 +1,43 @@
+        sleep(1) # wait for ajax
+
+        expect(click_mock).to have_received(:log)
+      end
```
You are right, testing for the end output of a logline is better. I got this working, it took a long time to set up the correct assertion while still having the logline written. Done though.

ondrae commented:
@MothOnMars My last few commits have been trying different approaches for the `invalid byte sequence` in VCR errors. The only one that passes CI is the [setting up](https://github.com/GSA/search-gov/pull/583/commits/81825902fd53f857445276e1bf9452c8583e9c42) that specific test.

MothOnMars commented on spec/fixtures/affiliates.yml:
```diff
@@ -5,6 +5,13 @@ DEFAULTS: &DEFAULTS
+bing_v7_affiliate:
+  <<: *DEFAULTS
+  search_engine: BingV7
+  display_name: Bing7 Affiliate
+  name: bing7affiliate
```
Let's use the "v" consistently:
```
display_name: Bing V7 Affiliate
name: bingV7affiliate
```

MothOnMars commented on spec/fixtures/affiliates.yml:
```diff
@@ -5,6 +5,13 @@ DEFAULTS: &DEFAULTS
+  search_engine: BingV7
+  display_name: Bing7 Affiliate
+  name: bing7affiliate
+  active: true
```
`active` defaults to `true`, so this should not need to be set in the fixtures:
```
  `active` tinyint(1) NOT NULL DEFAULT '1',
```

MothOnMars commented on spec/helpers/results_helper_spec.rb:
```diff
@@ -0,0 +1,37 @@
+             query: 'rutabaga')
+    end
+
+    subject { search_data(search, 'i14y') }
+
+    it 'adds data attributes to #search needed for click tracking' do
+      expected_output = {
+        data: {
+          affiliate: 'nps.gov',
+          vertical: 'i14y',
+          query: 'rutabaga'
+        }
+      }
+
+      expect(subject).to eq expected_output
+    end
+  end
+
+  describe '#link_to_result_title' do
+    subject { link_to_result_title('test title', 'https://test.gov', '2', 'BOOS') }
+
+    it 'makes a link with the added data-click attribute' do
+      expected_output = '<a data-click="{&quot;position&quot;:&quot;2&quot;,'\
+                        '&quot;module_code&quot;:&quot;BOOS&quot;}"'\
+                        ' href="https://test.gov">test title</a>'
+
+      expect(subject).to eq expected_output
+    end
+  end
+end
```
You can use a named subject in these specs:
```ruby
    subject(:search_data) { helper.search_data(search, 'i14y') }

    it 'adds data attributes to #search needed for click tracking' do
  ...
      expect(search_data).to eq expected_output
    end
  end
```

ondrae commented:
Adding `# frozen_string_literal: true` to either of the click request specs causes the below error. I'd like to ignore these two specific CodeClimate suggestions. I'm not sure how to do that though.
```
Failure/Error: CGI::unescape(env['REQUEST_URI'].force_encoding('UTF-8'))
     
     FrozenError:
       can't modify frozen String

     # ./lib/middlewares/reject_invalid_request_uri.rb:9:in `force_encoding'
```

ondrae commented:
@MothOnMars I added three more commits.

https://github.com/GSA/search-gov/pull/583/commits/9353628b1398877a052998c3c954143c4255cfd4 in the test environment, CSRF token checking is turned off. We are using a POST in the API, so need to disable checking for that specific controller.

https://github.com/GSA/search-gov/pull/583/commits/0fc0d821a3e864135fda021ca95c0168266a081f Using the instance values worked at first. As we added more complicated validations, we added some instance variables, that we don't need to report to the outgoing logline. This commit makes a specific list of what to log, more like how it used to work.

https://github.com/GSA/search-gov/pull/583/commits/950f64cb3968f7e7eea48fe331068cff7b36a84f is just clearer naming.

ondrae commented:
@MothOnMars Ready for re-review. Changes since you last reviewed:
- You pushed the working url escaping and encoding code.
- Stopped checking for CSRF tokens on the click tracking POSTs, with specs
- All the methods in results_helper now create the right html attributes for click tracking to work on all the modules, with specs.
